rust/arrow/src/array/union.rs

```diff
@@ -0,0 +1,1174 @@
+        }
+
+        for (child_index, name) in self.type_names().iter().enumerate() {
+            let column = &self.boxed_fields[child_index];
+            write!(
+                f,
+                "-- child {}: \"{}\" ({:?})\n",
+                child_index,
+                *name,
+                column.data_type()
+            )?;
+            fmt::Debug::fmt(column, f)?;
+            write!(f, "\n")?;
+        }
+        write!(f, "]")
+    }
+}
+
+/// `FieldData` is a helper struct to track the state of the fields in the `UnionBuilder`.
+struct FieldData {
+    /// The type id for this field
+    type_id: i8,
+    /// The Arrow data type represented in the `values_buffer`, which is untyped
+    data_type: DataType,
+    /// A buffer containing the values for this field in raw bytes
+    values_buffer: Option<MutableBuffer>,
+    ///  The number of array slots represented by the buffer
+    slots: usize,
+    /// The number of null array slots in this child array
+    null_count: usize,
+    /// A builder for the bitmap if required
+    bitmap_builder: Option<BooleanBufferBuilder>,
+}
+
+impl FieldData {
+    /// Creates a new `FieldData`.
+    fn new(
+        type_id: i8,
+        data_type: DataType,
+        bitmap_builder: Option<BooleanBufferBuilder>,
+    ) -> Self {
+        Self {
+            type_id,
+            data_type,
+            // TODO: Should `MutableBuffer` implement `Default`?
+            values_buffer: Some(MutableBuffer::new(1)),
+            slots: 0,
+            null_count: 0,
+            bitmap_builder,
+        }
+    }
+
+    /// Appends a single value to this `FieldData`'s `values_buffer`.
+    fn append_to_values_buffer<T: ArrowPrimitiveType>(
+        &mut self,
+        v: T::Native,
+    ) -> Result<()> {
+        let values_buffer = self
+            .values_buffer
+            .take()
+            .expect("Values buffer was never created");
+        let mut builder: BufferBuilder<T> =
+            mutable_buffer_to_builder(values_buffer, self.slots);
+        builder.append(v)?;
+        let mutable_buffer = builder_to_mutable_buffer(builder);
+        self.values_buffer = Some(mutable_buffer);
+
+        self.slots += 1;
+        if let Some(b) = &mut self.bitmap_builder {
+            b.append(true)?
+        };
+        Ok(())
+    }
+
+    /// Appends a null to this `FieldData`.
+    fn append_null<T: ArrowPrimitiveType>(&mut self) -> Result<()> {
+        if let Some(b) = &mut self.bitmap_builder {
+            let values_buffer = self
+                .values_buffer
+                .take()
+                .expect("Values buffer was never created");
+            let mut builder: BufferBuilder<T> =
+                mutable_buffer_to_builder(values_buffer, self.slots);
+            builder.advance(1)?;
+            let mutable_buffer = builder_to_mutable_buffer(builder);
+            self.values_buffer = Some(mutable_buffer);
+            self.slots += 1;
```

Comment:
Do we increment `self.slots` or `self.null_count` when appending null?

Reply:
Actually I should have incremented both.  The `null_count` of the `ArrayData` for the children was not set correctly.  Will update.